hardware_interface/test/test_component_interfaces.cpp

```diff
@@ -0,0 +1,577 @@
+
+using namespace ::testing;  // NOLINT
+
+namespace hardware_interface
+{
+
+namespace hardware_interfaces_components_test
+{
+
+class DummyPositionJoint : public Joint
+{
+public:
+  return_type configure(const ComponentInfo & joint_info) override
+  {
+    info = joint_info;
+    if (info.command_interfaces.size() > 1 || info.state_interfaces.size() > 1) {
+      return return_type::ERROR;
+    }
+    if (info.command_interfaces.size() == 0) {
+      info.command_interfaces.push_back("position");
+    }
+    if (info.state_interfaces.size() == 0) {
+      info.state_interfaces.push_back("position");
+    }
+    max_position = stod(info.parameters["max_position"]);
+    min_position = stod(info.parameters["min_position"]);
+    return return_type::OK;
+  }
+
+  std::vector<std::string> get_command_interfaces() const override
+  {
+    return info.command_interfaces;
+  }
+
+  std::vector<std::string> get_state_interfaces() const override
+  {
+    return info.state_interfaces;
+  }
+
+  return_type set_command(
+    const std::vector<double> command,
+    std::vector<std::string> interfaces = std::vector<std::string>()) override
+  {
+    if (interfaces.size() != 0) {
+      return return_type::ERROR;
+    }
+    position_command = command[0];
+    return return_type::OK;
+  }
+
+  return_type get_command(
+    std::vector<double> & command,
+    std::vector<std::string> & interfaces) const override
+  {
+    command.push_back(position_command);
+    interfaces = info.command_interfaces;
+    return return_type::OK;
+  }
+
+  return_type set_state(
+    const std::vector<double> & state,
+    std::vector<std::string> interfaces = std::vector<std::string>()) override
+  {
+    return_type ret = return_type::OK;
+    if (interfaces.size() != 0) {
+      ret = return_type::ERROR;
+    }
+    if (state[0] > min_position && state[0] < max_position) {
+      position_state = state[0];
+    } else {
+      ret = return_type::ERROR;
+    }
+    return ret;
+  }
+
+  return_type get_state(
+    std::vector<double> & state,
+    std::vector<std::string> & interfaces) const override
+  {
+    state.push_back(position_state);
+    interfaces = info.command_interfaces;
+    return return_type::OK;
+  }
+
+private:
+  ComponentInfo info;
+  double position_command;
+  double position_state;
+  double max_position, min_position;
+};
+
+class DummyForceTorqueSensor : public Sensor
+{
+public:
+  return_type configure(const ComponentInfo & sensor_info) override
+  {
+    info = sensor_info;
+    if (info.parameters["frame_id"] == "") {
+      return return_type::ERROR;
+    }
+    if (info.state_interfaces.size() == 0) {
+      info.state_interfaces.push_back("force_x");
+      info.state_interfaces.push_back("force_y");
+      info.state_interfaces.push_back("force_z");
+      info.state_interfaces.push_back("torque_x");
+      info.state_interfaces.push_back("torque_y");
+      info.state_interfaces.push_back("torque_z");
+    }
+    ft_values.resize(6);
```

Comment:
nitpick: no need to resize here if the copy assignment operator is being used later on.